### ISSUE TITLE
🐛 BUG: Snowpack fails to find `react-dom` when building `examples/docs`

### ISSUE THREAD

**Issue by togami2864** [snowpackjs/astro#1311]
### What package manager are you using?

npm v7.21.0

### What operating system are you using?

Mac 

### Describe the Bug

When I select the template <Documentation>and run build, it doesn't work well for building react-dom@16.14.0.

## avoid the error
I run the command `npm ls react-dom` and then got the following
```
@example/docs@0.0.1 /Users/a/Desktop/astro_docs
├─┬@docsearch/react@1.0.0-alpha.28
│ └── react-dom@16.14.0
└─┬ astro@0.20.4
  └─┬ @astrojs/renderer-react@0.2.0
    └── react-dom@17.0.2
```
@docsearch/react@1.0.0-alpha.28 depends on react-dom@16.14.0.

So, I did

1: npm uninstall @docsearch/react
2: npm install @docsearch/react@alpha <- (install @docsearch/react@3.0.0-alpha.40)
3: npm run dev
4: success !!!

If necessary, I'll open a PR to version up the package.json.
(In addition, oddly enough, even though the latest version (v3) of @docsearch/react has been released, the version on [the npm page](https://www.npmjs.com/package/@docsearch/react) is stuck at v1.)

### Steps to Reproduce

1. `npm init astro` using template <Documentation>
2. npm install
3. npm run dev
4. display error as the following
```
$ npm run dev

> @example/docs@0.0.1 dev
> astro dev

[22:47:56] [snowpack] Hint: run "snowpack init" to create a project config file. Using defaults...
[22:47:56] [snowpack] Welcome to Snowpack! Because this is your first time running
this project, Snowpack needs to prepare your dependencies. This is a one-time step
and the results will be cached for the lifetime of your project. Please wait...
[22:47:56] [snowpack] + astro/dist/internal/__astro_component.js@0.20.4
[22:47:57] [snowpack] └── shorthash@0.0.2 (dedupe)
[22:47:57] [snowpack] └── estree-util-value-to-estree@1.2.0 (dedupe)
[22:47:58] [snowpack]   └── is-plain-obj@3.0.0
[22:47:58] [snowpack] └── astring@1.7.5 (dedupe)
[22:47:58] [snowpack] + astro/dist/internal/element-registry.js@0.20.4
[22:47:58] [snowpack] + astro/dist/internal/fetch-content.js@0.20.4
[22:47:59] [snowpack] + astro/dist/internal/__astro_slot.js@0.20.4
[22:47:59] [snowpack] + astro/dist/internal/__astro_hoisted_scripts.js@0.20.4
[22:48:00] [snowpack] + prismjs@1.24.1
[22:48:00] [snowpack] + @astrojs/renderer-preact/server@0.2.1
[22:48:01] [snowpack] └── preact@10.5.14 (dedupe)
[22:48:01] [snowpack] └── preact-render-to-string@5.1.19 (dedupe)
[22:48:01] [snowpack] + @astrojs/renderer-preact/client@0.2.1
[22:48:01] [snowpack] + preact/jsx-runtime@10.5.14
[22:48:02] [snowpack] + @astrojs/renderer-react/server@0.2.0
[22:48:02] [snowpack] └── react@17.0.2 (dedupe)
[22:48:02] [snowpack]   └── object-assign@4.1.1
[22:48:02] [snowpack] └── react-dom/server.js@17.0.2 (dedupe)
[22:48:03] [snowpack] + @astrojs/renderer-react/client@0.2.0
[22:48:03] [snowpack] └── react-dom@17.0.2 (dedupe)
[22:48:07] [snowpack]   └── scheduler@0.20.2
[22:48:07] [snowpack]   └── scheduler/tracing@0.20.2
[22:48:07] [snowpack] + react@16.14.0
[22:48:07] [snowpack] └── prop-types/checkPropTypes@15.7.2
[22:48:07] [snowpack] + react/jsx-runtime@16.14.0
[22:48:08] [snowpack] + react-dom@16.14.0
node:internal/process/promises:246
          triggerUncaughtException(err, true /* fromPromise */);
          ^

[Error: ENOENT: no such file or directory, open '/Users/a/Desktop/astro_docs/node_modules/.cache/snowpack/build/react-dom@16.14.0/react-dom/server.js'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/Users/a/Desktop/astro_docs/node_modules/.cache/snowpack/build/react-dom@16.14.0/react-dom/server.js'
}
```

**Comment by jasikpark** [comment]
Seems to build on https://codesandbox.io/s/github/snowpackjs/astro/tree/main/examples/docs, but this is two issues with the same result... 

Can you try running that with `npm run dev --verbose` and copying the output?

**Comment by MattWilcox** [comment]
Copying over an error log from [another bug](https://github.com/snowpackjs/astro/issues/1325) that seems to be the same issue as this...

First failure:
```
➜ npm init astro -- --template docs

Need to install the following packages:
  create-astro
Ok to proceed? (y) y

Welcome to Astro! (create-astro v0.5.2)
If you encounter a problem, visit https://github.com/snowpackjs/astro/issues to search or file a new issue.

> Prepare for liftoff.
> Gathering mission details...
✔ Using template's default renderers › Preact
> Copying project files...
✔ Done!

Next steps:
  1: npm install (or pnpm install, yarn, etc)
  2: git init && git add -A && git commit -m "Initial commit" (optional step)
  3: npm start (or pnpm, yarn, etc)

To close the dev server, hit Ctrl-C

Stuck? Visit us at https://astro.build/chat


nitro/astro/docs is 📦 v0.0.1 via ⬢ v14.17.0 took 6s 
➜ npm install
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated fsevents@2.1.3: "Please update to latest v2.3 or v2.2"
npm WARN deprecated fsevents@2.1.3: "Please update to latest v2.3 or v2.2"
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142

added 827 packages, and audited 828 packages in 44s

190 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

nitro/astro/docs is 📦 v0.0.1 via ⬢ v14.17.0 took 44s 
➜ npm run dev                      

> @example/docs@0.0.1 dev
> astro dev

[21:08:22] [snowpack] Hint: run "snowpack init" to create a project config file. Using defaults...
[21:08:22] [snowpack] Welcome to Snowpack! Because this is your first time running
this project, Snowpack needs to prepare your dependencies. This is a one-time step
and the results will be cached for the lifetime of your project. Please wait...
[21:08:23] [snowpack] + astro/dist/internal/__astro_component.js@0.20.4
[21:08:23] [snowpack] └── shorthash@0.0.2 (dedupe)
[21:08:23] [snowpack] └── estree-util-value-to-estree@1.2.0 (dedupe)
[21:08:23] [snowpack]   └── is-plain-obj@3.0.0
[21:08:23] [snowpack] └── astring@1.7.5 (dedupe)
[21:08:24] [snowpack] + astro/dist/internal/element-registry.js@0.20.4
[21:08:24] [snowpack] + astro/dist/internal/fetch-content.js@0.20.4
[21:08:24] [snowpack] + astro/dist/internal/__astro_slot.js@0.20.4
[21:08:24] [snowpack] + astro/dist/internal/__astro_hoisted_scripts.js@0.20.4
[21:08:24] [snowpack] + prismjs@1.24.1
[21:08:25] [snowpack] + @astrojs/renderer-preact/server@0.2.1
[21:08:25] [snowpack] └── preact@10.5.14 (dedupe)
[21:08:25] [snowpack] └── preact-render-to-string@5.1.19 (dedupe)
[21:08:25] [snowpack] + @astrojs/renderer-preact/client@0.2.1
[21:08:25] [snowpack] + preact/jsx-runtime@10.5.14
[21:08:25] [snowpack] + @astrojs/renderer-react/server@0.2.0
[21:08:25] [snowpack] └── react@17.0.2 (dedupe)
[21:08:26] [snowpack]   └── object-assign@4.1.1
[21:08:26] [snowpack] └── react-dom/server.js@17.0.2 (dedupe)
[21:08:26] [snowpack] + @astrojs/renderer-react/client@0.2.0
[21:08:26] [snowpack] └── react-dom@17.0.2 (dedupe)
[21:08:28] [snowpack]   └── scheduler@0.20.2
[21:08:28] [snowpack]   └── scheduler/tracing@0.20.2
[21:08:28] [snowpack] + react@16.14.0
[21:08:28] [snowpack] └── prop-types/checkPropTypes@15.7.2
[21:08:28] [snowpack] + react/jsx-runtime@16.14.0
[21:08:29] [snowpack] + react-dom@16.14.0
(node:70918) UnhandledPromiseRejectionWarning: Error: ENOENT: no such file or directory, open '/Users/mattwilcox/nitro/astro/docs/node_modules/.cache/snowpack/build/react-dom@16.14.0/react-dom/server.js'
(Use `node --trace-warnings ...` to show where the warning was created)
(node:70918) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:70918) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

nitro/astro/docs is 📦 v0.0.1 via ⬢ v14.17.0 took 9s 
```

Second failure:

```➜ npm install
npm WARN read-shrinkwrap This version of npm is compatible with lockfileVersion@1, but package-lock.json was generated for lockfileVersion@2. I'll try to do my best with it!

> bufferutil@4.0.3 install /Users/mattwilcox/nitro/astro/docs/node_modules/bufferutil
> node-gyp-build


> utf-8-validate@5.0.5 install /Users/mattwilcox/nitro/astro/docs/node_modules/utf-8-validate
> node-gyp-build


> esbuild@0.12.25 postinstall /Users/mattwilcox/nitro/astro/docs/node_modules/esbuild
> node install.js


> esbuild@0.9.7 postinstall /Users/mattwilcox/nitro/astro/docs/node_modules/snowpack/node_modules/esbuild
> node install.js


> svelte-preprocess@4.9.1 postinstall /Users/mattwilcox/nitro/astro/docs/node_modules/svelte-preprocess
> echo "[svelte-preprocess] Don't forget to install the preprocessors packages that will be used: node-sass/sass, stylus, less, postcss & postcss-load-config, coffeescript, pug, etc..."

[svelte-preprocess] Don't forget to install the preprocessors packages that will be used: node-sass/sass, stylus, less, postcss & postcss-load-config, coffeescript, pug, etc...
removed 5 packages, updated 822 packages and audited 823 packages in 10.319s

190 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities


nitro/astro/docs is 📦 v0.0.1 via ⬢ v14.17.6 took 11s 
➜ npm run dev

> @example/docs@0.0.1 dev /Users/mattwilcox/nitro/astro/docs
> astro dev

[11:36:46] [snowpack] Hint: run "snowpack init" to create a project config file. Using defaults...
[11:36:46] [snowpack] Welcome to Snowpack! Because this is your first time running
this project, Snowpack needs to prepare your dependencies. This is a one-time step
and the results will be cached for the lifetime of your project. Please wait...
(node:74952) UnhandledPromiseRejectionWarning: Error: Package "react" not found. Have you installed it? 
    at Object.resolveEntrypoint (/Users/mattwilcox/nitro/astro/docs/node_modules/esinstall/lib/entrypoints.js:175:15)
    at PackageSourceLocal.buildPackageImport (/Users/mattwilcox/nitro/astro/docs/node_modules/snowpack/lib/cjs/sources/local.js:424:40)
    at async PackageSourceLocal.prepare (/Users/mattwilcox/nitro/astro/docs/node_modules/snowpack/lib/cjs/sources/local.js:227:13)
    at async startServer (/Users/mattwilcox/nitro/astro/docs/node_modules/snowpack/lib/cjs/commands/dev.js:240:9)
    at async createSnowpack (file:///Users/mattwilcox/nitro/astro/docs/node_modules/astro/dist/runtime.js:273:14)
    at async createRuntime (file:///Users/mattwilcox/nitro/astro/docs/node_modules/astro/dist/runtime.js:293:7)
    at async dev (file:///Users/mattwilcox/nitro/astro/docs/node_modules/astro/dist/dev.js:17:19)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:74952) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:74952) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by MattWilcox** [comment]
I don't know enough about SnowPack to be sure, but as a heads up I *have* had issues with SkyPack before where the CDNs crap out and do not deliver the correct payload for the requested URI. Its an intermittent problem that stopped me using SkyPack, but if there's a similar problem with SnowPack's resolution mechanic, that might explain how the same commands and projects run most of the time, but not all of the time; and for some people but not others if they're pulling from a different CDN location.

https://github.com/skypackjs/skypack-cdn/issues/168

**Comment by togami2864** [comment]
@jasikpark 
This is output of `npm run dev --verbose`
```
$ npm run dev --verbose
npm verb cli [
npm verb cli   '/Users/a/.nodenv/versions/16.8.0/bin/node',
npm verb cli   '/Users/a/.nodenv/versions/16.8.0/bin/npm',
npm verb cli   'run',
npm verb cli   'dev',
npm verb cli   '--verbose'
npm verb cli ]
npm info using npm@7.21.0
npm info using node@v16.8.0
npm timing npm:load:whichnode Completed in 0ms
npm timing config:load:defaults Completed in 2ms
npm timing config:load:file:/Users/a/.nodenv/versions/16.8.0/lib/node_modules/npm/npmrc Completed in 0ms
npm timing config:load:builtin Completed in 0ms
npm timing config:load:cli Completed in 3ms
npm timing config:load:env Completed in 0ms
npm timing config:load:file:/Users/a/Desktop/astro_doc/.npmrc Completed in 2ms
npm timing config:load:project Completed in 4ms
npm timing config:load:file:/Users/a/.npmrc Completed in 1ms
npm timing config:load:user Completed in 1ms
npm timing config:load:file:/Users/a/.nodenv/versions/16.8.0/etc/npmrc Completed in 1ms
npm timing config:load:global Completed in 1ms
npm timing config:load:validate Completed in 0ms
npm timing config:load:credentials Completed in 1ms
npm timing config:load:setEnvs Completed in 1ms
npm timing config:load Completed in 13ms
npm timing npm:load:configload Completed in 13ms
npm timing npm:load:setTitle Completed in 20ms
npm timing npm:load:setupLog Completed in 0ms
npm timing config:load:flatten Completed in 3ms
npm timing npm:load:cleanupLog Completed in 2ms
npm timing npm:load:configScope Completed in 0ms
npm timing npm:load:projectScope Completed in 1ms
npm timing npm:load Completed in 44ms
⸨⠂⠂⠂⠂⠂⠂⠂⠂⠂⠂⠂⠂⠂⠂⠂⠂⠂⠂⸩ ⠧ : timing npm:load Completed in 44ms
> @example/docs@0.0.1 dev
> astro dev

[19:40:25] [snowpack] Hint: run "snowpack init" to create a project config file. Using defaults...
[19:40:25] [snowpack] Welcome to Snowpack! Because this is your first time running
this project, Snowpack needs to prepare your dependencies. This is a one-time step
and the results will be cached for the lifetime of your project. Please wait...
[19:40:25] [snowpack] + astro/dist/internal/__astro_component.js@0.20.4
[19:40:27] [snowpack] └── shorthash@0.0.2 (dedupe) in 44ms
[19:40:27] [snowpack] └── estree-util-value-to-estree@1.2.0 (dedupe)
[19:40:27] [snowpack]   └── is-plain-obj@3.0.0eted in 44ms
[19:40:28] [snowpack] └── astring@1.7.5 (dedupe)ed in 44ms
[19:40:28] [snowpack] + astro/dist/internal/element-registry.js@0.20.4
[19:40:28] [snowpack] + astro/dist/internal/fetch-content.js@0.20.4
[19:40:29] [snowpack] + astro/dist/internal/__astro_slot.js@0.20.4
[19:40:29] [snowpack] + astro/dist/internal/__astro_hoisted_scripts.js@0.20.4
[19:40:30] [snowpack] + prismjs@1.24.1ad Completed in 44ms
[19:40:30] [snowpack] + @astrojs/renderer-preact/server@0.2.1
[19:40:30] [snowpack] └── preact@10.5.14 (dedupe)d in 44ms
[19:40:31] [snowpack] └── preact-render-to-string@5.1.19 (dedupe)
[19:40:31] [snowpack] + @astrojs/renderer-preact/client@0.2.1
[19:40:31] [snowpack] + preact/jsx-runtime@10.5.14 in 44ms
[19:40:31] [snowpack] + @astrojs/renderer-react/server@0.2.0
[19:40:31] [snowpack] └── react@17.0.2 (dedupe)ted in 44ms
[19:40:32] [snowpack]   └── object-assign@4.1.1ted in 44ms
[19:40:32] [snowpack] └── react-dom/server.js@17.0.2 (dedupe)
[19:40:33] [snowpack] + @astrojs/renderer-react/client@0.2.0
[19:40:33] [snowpack] └── react-dom@17.0.2 (dedupe)in 44ms
[19:40:37] [snowpack]   └── scheduler@0.20.2pleted in 44ms
[19:40:37] [snowpack]   └── scheduler/tracing@0.20.2n 44ms
[19:40:37] [snowpack] + react@16.14.0oad Completed in 44ms
[19:40:38] [snowpack] └── prop-types/checkPropTypes@15.7.2
[19:40:38] [snowpack] + react/jsx-runtime@16.14.0d in 44ms
[19:40:38] [snowpack] + react-dom@16.14.0Completed in 44ms
node:internal/process/promises:246m:load Completed in 44ms
          triggerUncaughtException(err, true /* fromPromise */);
          ^

[Error: ENOENT: no such file or directory, open '/Users/a/Desktop/astro_doc/node_modules/.cache/snowpack/build/react-dom@16.14.0/react-dom/server.js'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/Users/a/Desktop/astro_doc/node_modules/.cache/snowpack/build/react-dom@16.14.0/react-dom/server.js'
}
npm timing command:run-script Completed in 18237ms
npm verb exit 1
npm timing npm Completed in 18615ms
npm verb code 1
```

**Comment by jasikpark** [comment]
I suspect this could be a snowpack problem? 🤔

**Comment by MattWilcox** [comment]
Pretty sure it's a snowpack problem and specifically Snowpack failing to get the requested file and not throwing a useful error when the file doesn't download to cache. Which is a similar issue I've had with skypack. It's a case of... why?

**Comment by huphtur** [comment]
Can confirm same error, tried multiple Node versions.

**Comment by FredKSchott** [comment]
Thanks for the report! We believe that this has been fixed in astro v0.21, going out today. Let us know if you can still reproduce the issue after updating to v0.21 and we'll be happy to take a look.
